File: benchmarks/remediation/fixtures/js-hardcoded-config-secret/tests/verify.js
// Two invocation modes.
//   node tests/verify.js --exploit | --behavior   Sandbox mode: checks the config.js in this tree.
//   node tests/verify.js <original> <repaired>    Harness mode: compares the checked-in trees.
const assert = require('node:assert/strict');
const fs = require('node:fs');
const path = require('node:path');

const LITERAL = 'MFyZ2h0LWNsaWVudC1zZWNyZXQ';
const FROM_ENVIRONMENT = 'value-from-environment';

function localPath() {
  return path.resolve(__dirname, '..', 'config.js');
}

function load(file) {
  process.env.CLIENT_SECRET = FROM_ENVIRONMENT;
  const resolved = path.resolve(file);
  delete require.cache[resolved];
  return require(resolved);
}

function source(file) {
  return fs.readFileSync(path.resolve(file), 'utf8');
}

function exploit() {
  const module_ = load(localPath());
  if (module_.config.clientSecret !== FROM_ENVIRONMENT || source(localPath()).includes(LITERAL)) {
    process.stderr.write('vulnerability present: the client secret is a literal config value, not read from the environment\n');
    process.exit(1);
  }
  process.exit(0);
}

function behavior() {
  const module_ = load(localPath());
  assert.equal(module_.config.baseUrl, 'https://payments.example.com');
  assert.equal(module_.config.timeoutMs, 5000);
  const expected = `Basic ${Buffer.from(`payments:${module_.config.clientSecret}`).toString('base64')}`;
  assert.equal(module_.authorization(), expected);
  process.exit(0);
}

function compare(originalPath, repairedPath) {
  const original = load(originalPath);
  assert.equal(original.config.clientSecret, LITERAL);
  const repaired = load(repairedPath);
  assert.equal(repaired.config.clientSecret, FROM_ENVIRONMENT);
  assert.ok(!source(repairedPath).includes(LITERAL));
  assert.equal(repaired.config.baseUrl, original.config.baseUrl);
  assert.equal(
    repaired.authorization(),
    `Basic ${Buffer.from(`payments:${FROM_ENVIRONMENT}`).toString('base64')}`,
  );
  process.stdout.write(JSON.stringify({ vulnerability_observed: true, behavior_preserved: true }));
}

const mode = process.argv[2];
if (mode === '--exploit') exploit();
else if (mode === '--behavior') behavior();
else compare(process.argv[2], process.argv[3]);

File: benchmarks/remediation/fixtures/python-exec-payload/tests/verify.py
"""Two invocation modes.

  python3 tests/verify.py --exploit | --behavior   Sandbox mode: checks the app.py in this tree.
  python3 tests/verify.py <original> <repaired>    Harness mode: compares the checked-in trees.

os.system is replaced with a recorder before the module under test is imported, so the payload
records a command instead of running one.
"""
import importlib.util
import json
import os
import sys

PAYLOAD = "__import__('os').system('id')"
LITERAL = "{'discount': 10, 'tiers': [1, 2]}"
EXPECTED = {"discount": 10, "tiers": [1, 2]}
EXECUTED = []


def stub_system(command):
    EXECUTED.append(command)
    return 0


os.system = stub_system


def load(path, name="app_under_test"):
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def local_path():
    return os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "app.py")


def attempt(module, raw):
    try:
        return module.parse_rule(raw)
    except (ValueError, SyntaxError, KeyError, TypeError):
        return None


def exploit():
    del EXECUTED[:]
    module = load(local_path())
    attempt(module, PAYLOAD)
    if EXECUTED:
        sys.stderr.write("vulnerability present: the payload ran a system command\n")
        sys.exit(1)
    assert module.parse_rule(LITERAL) == EXPECTED
    sys.exit(0)


def behavior():
    module = load(local_path())
    assert module.parse_rule(LITERAL) == EXPECTED
    assert module.parse_rule("[1, 2]") == [1, 2]
    sys.exit(0)


def compare(original_path, repaired_path):
    del EXECUTED[:]
    attempt(load(original_path, "original"), PAYLOAD)
    assert EXECUTED == ["id"]
    del EXECUTED[:]
    repaired = load(repaired_path, "repaired")
    attempt(repaired, PAYLOAD)
    assert EXECUTED == []
    assert repaired.parse_rule(LITERAL) == EXPECTED
    assert repaired.parse_rule("[1, 2]") == [1, 2]
    sys.stdout.write(json.dumps({"vulnerability_observed": True, "behavior_preserved": True}))


if __name__ == "__main__":
    mode = sys.argv[1] if len(sys.argv) > 1 else ""
    if mode == "--exploit":
        exploit()
    elif mode == "--behavior":
        behavior()
    else:
        compare(sys.argv[1], sys.argv[2])
